File: 剑指OFFER/剑指 Offer 64. 求1+2+…+n/sumNums.go
package main

import (
	"fmt"
)

func main() {
	n1 := 3
	n2 := 9
	fmt.Println(n1, n2)
	fmt.Println(sumNums(n1),sumNums(n2))
}

func sumNums(n int) int {
	ans := 0
	var sum func(n int) bool
	sum = func(n int) bool {
		ans += n
		return n > 0 && sum(n-1)
	}
	sum(n)
	return ans
}

File: 剑指OFFER/剑指 Offer 64. 求1+2+…+n/sumNums.php
<?php

class Solution {

    /**
     * @param Integer $n
     * @return Integer
     */
    public $ans = 0;
    function sumNums($n) {

        $this->ans += $n;
        $n > 0 && $this->sumNums($n-1);
        return $this->ans;

    }
}

$s = new Solution();
var_dump($s->sumNums(3));

//不使用类的方法
global $ans;
$ans = 0;
function sumNums($n) {

    $GLOBAL["ans"]+=$n;
    echo $n.' '.$GLOBAL["ans"].PHP_EOL;//不使用全局变量的话，会造成闭包问题。
    $n > 0 && sumNums($n-1);
    return $GLOBAL["ans"];
}

var_dump(sumNums(3));
